Add ERC: Wallet Capabilities for Account Abstraction (#947)
ethereum/ERCs: ERCS/erc-xxxx.md

---
eip:
title: Wallet Capabilities for Account Abstraction
description: New Wallet Capabilities providing a way for dApps and Account Abstraction wallets to communicate
author: Yoav Weiss (@yoavw), Alex Forshtat (@forshtat), Dror Tirosh (@drortirosh), Shahaf Nacson (@shahafn)
discussions-to:
status: Draft
type: Standards Track
category: ERC
created:
requires: 5792, 7702
---

## Abstract

[EIP-5792](./eip-5792) defines a baseline JSON-RPC API for a communication between wallets and dapps.

> With EIP-5792, apps and wallets can communicate about any advanced features using "capabilities" - extensions
> to the base protocol that must be defined in separate documents.

Reviewer: ```suggestion
With EIP-5792, apps and wallets can communicate about any advanced features using "capabilities" - extensions to the base protocol that must be defined in separate documents.
```

> This proposal defines a set of "capabilities" the wallets may want to implement in order to provide a
> comprehensive support for Account Abstraction.

Reviewer 2: ```suggestion
This proposal defines a set of "capabilities" the wallets may want to implement in order to provide a comprehensive support for Account Abstraction.
```

## Motivation

> [ERC-4337](./eip-4337.md) introduced Account Abstraction and enabling Smart Contract Accounts to function as first-class citizens in Ethereum.

Reviewer 2: ```suggestion
[ERC-4337](./eip-4337.md) introduced Account Abstraction, enabling Smart Contract Accounts to function as first-class citizens in Ethereum.
```

However, while [ERC-4337](./eip-4337.md) and [ERC-7769](./eip-7769.md) defines a low-level RPC API for Account Abstraction,
it does not specify a way for advanced AA-aware dApps to communicate their supported features and parameters to the advanced AA Wallet Applications.

> This ERC addresses the issue by defining a structured set of capabilities tailored for Account Abstraction dApps and Wallet Applications.

Reviewer 2: ```suggestion
This ERC addresses the issue by defining a structured set of capabilities tailored for AA-aware dApps and Wallet Applications.
```

> It extends the [EIP-5792](./eip-5792) wallet capability model to encompass critical aspects of AA,

Author: ```suggestion
It utilises the [EIP-5792](./eip-5792) wallet capability model to express some of the critical aspects of AA,
```

> It extends the [EIP-5792](./eip-5792) wallet capability model to encompass critical aspects of AA,
> ensuring dApps can seamlessly adapt to different AA Wallets without requiring custom solutions.

Reviewer: ```suggestion
It extends the [EIP-5792](./eip-5792) wallet capability model to encompass critical aspects of AA, ensuring dApps can seamlessly adapt to different AA Wallets without requiring custom solutions.
```

> ensuring dApps can seamlessly adapt to different AA Wallets without requiring custom solutions.

Reviewer 2: Is it an extension, or 5792 capabilities proposals?

Author: Rephrased

## Specification

All actions in Account Abstraction within the context of EIP-5792 must be done on a single chain and atomically.

We define the following list of new "capabilities" which together cover many features necessary for Account Abstraction.
Note that use of Paymasters managed by a "paymaster web service" is described in [ERC-7677](./eip-7677).

### Create [EIP-7702](./eip-7702) Authorization Capability

> This capability is designed to use with [EIP-7702](./eip-7702) and requests the Wallet Application to provide

Reviewer 2: ```suggestion
This capability is designed to be used with [EIP-7702](./eip-7702) and requests the Wallet Application to provide
```

> This capability is designed to use with [EIP-7702](./eip-7702) and requests the Wallet Application to provide
> an EIP-7702 authorization tuple for the specified address as part of the AA transaction.

Reviewer: ```suggestion
This capability is designed to use with [EIP-7702](./eip-7702) and requests the Wallet Application to provide an EIP-7702 authorization tuple for the specified address as part of the AA transaction.
```

Identifier:

`eip7702Auth`

Interface:

```typescript
type SetCodeForEOACapabilityParams = Record<
  {
    account: `0x${string}`,       // EOA address
    delegation: `0x${string}`,    // delegation address
  }
>
```

Supporting Wallet Applications MUST generate an EIP-7702 compatible transaction that sets a code of the `account` EOA address
to the code of `delegation` specified in the request.

Attention! Wallet Applications MUST maintain a strict shortlist of well-known and publicly audited Smart Contract Account

> implementations that can be acceptable as `delegation`.

Reviewer 2: ```suggestion
implementations that are acceptable as `delegation`.
```

Authorization is an extremely sensitive operation and any vulnerability or malicious code in `delegation` will
result in complete draining of the `account`.

### Static Paymaster Configuration Capability

The purpose of this capability is allowing applications to integrate with Paymasters that do not require
the Wallet Application to resolve any dynamic configuration.

The application may hard-code or resolve these parameters first and pass them with this capability.

Identifier:

`staticPaymasterConfiguration`

Interface:

```typescript
type StaticPaymasterConfigurationCapabilityParams = Record<
  {
    paymaster: string;
    paymasterData: string;
    paymasterValidationGasLimit: `0x${string}`;
    paymasterPostOpGasLimit: `0x${string}`;
  }
>;
```

### Validity Time Range Capability

> The purpose of this capability is allowing the applications to explicitly specify the time range during which
> the requested operations will be valid after signing.

Reviewer: ```suggestion
The purpose of this capability is allowing the applications to explicitly specify the time range during which the requested operations will be valid after signing.
```

Identifier:

`validityTimeRange`

Interface:

```typescript
type ValidityTimeRangeCapabilityParams = Record<
  {
    validAfter: `0x${string}`, // operation valid only after this timestamp, in seconds
    validUntil: `0x${string}`  // operation valid only before this timestamp, in seconds
  }
>
```

> The Wallet Application MUST verify the time range [`validAfter`..`validUntil`] is valid and present it to the
> user in a human-readable way for confirmation as part of the transaction information.

Reviewer: ```suggestion
The Wallet Application MUST verify the time range [`validAfter`..`validUntil`] is valid and present it to the user in a human-readable way for confirmation as part of the transaction information.
```

The Smart Contract Account MUST specify the time range [`validAfter`..`validUntil`] as the transaction validity range.

### Multidimensional Nonce Capability

> The purpose of this capability is allowing the applications to explicitly specify the components of the
> multidimensional nonce as defined in [ERC-4337](./eip-4337.md).

Reviewer: ```suggestion
The purpose of this capability is allowing the applications to explicitly specify the components of the multidimensional nonce as defined in [ERC-4337](./eip-4337.md).
```

> multidimensional nonce as defined in [ERC-4337](./eip-4337.md).

Author: ```suggestion
semi-abstracted nonce as defined in [ERC-4337](./eip-4337.md).
```

> multidimensional nonce as defined in [ERC-4337](./eip-4337.md).

Author: Fixed.
